### PR TITLE
[NCL-6166] Add build type flags for DA calls

### PR DESCRIPTION
**Pull request by pkocandr**
Introduce command line flag restBrewPullActive and new switch restMode. Those are used to determine if the REST service (DA) should use the Brew Pull feature or not and if it should also lookup temporary and/or managed service artifacts or not. This should replace passing of restRepositoryGroup value and the repository group to be used should be determined by DA based on the flags. All of those command line params are passed to the rest service which should ensure it's backwards compatible.

### CODE REVIEW

**Comment by pkocandr** [comment]
One known downside of this is that also the built projects modules' version is determined using these flags. In context of PNC the brewPullActive should always be true for them to ensure it doesn't rebuild an already existing version built in Brew. It would fail later in promotion validation, but it is too late.
To fix this in case restBrewPullActive=false is passed we should split the request to DA to check the latest versions of built modules with brewPullActive=true and versions of dependencies should used the passed value (brewPullActive=false).

File: core/src/main/java/org/commonjava/maven/ext/core/state/RESTState.java [hunk 1]
```diff
@@ -95,8 +105,9 @@ public void initialise( Properties userProps )
         int restRetryDuration = Integer.parseInt( userProps.getProperty( REST_RETRY_DURATION_SEC,
                                                                          String.valueOf( DefaultTranslator.RETRY_DURATION_SEC ) ) );
 
-        restEndpoint = new DefaultTranslator( restURL, restMaxSize, restMinSize, repositoryGroup, vState.getIncrementalSerialSuffix(),
-                                              restHeaders, restConnectionTimeout, restSocketTimeout, restRetryDuration );
+        restEndpoint = new DefaultTranslator( restURL, restMaxSize, restMinSize, repositoryGroup, brewPullActive, temporaryBuild,
```
Comment (rnc) [inline core/src/main/java/org/commonjava/maven/ext/core/state/RESTState.java:108]: Question - if `brewPullAction` and `temporaryBuild` are replacing `repositoryGroup` why do we still have `repositoryGroup` here?

**Comment by rnc** [comment]
Aside - I presume you would also be doing a matched PR for the documentation?

File: pom.xml [hunk 2]
```diff
@@ -78,6 +78,7 @@
 
     <mavenVersion>3.5.0</mavenVersion>
     <atlasVersion>0.17.1</atlasVersion>
+    <daVersion>2.1.0-SNAPSHOT</daVersion>
```
Comment (rnc) [inline pom.xml:81]: Is there not a non-snapshot version to use?

File: core/src/main/java/org/commonjava/maven/ext/core/state/RESTState.java (hunk 1 above)
Comment (pkocandr) [reply core/src/main/java/org/commonjava/maven/ext/core/state/RESTState.java:108]: For backwards compatibility so we can deploy new PME independently on DA/Repour. We can either use the new PME version earlier than them or later.

File: pom.xml (hunk 2 above)
Comment (pkocandr) [reply pom.xml:81]: Not yet. I was thinking about separating the API from DA and release it once finalized for PNC 2.1 even though we will be still in development phase. Have to discuss that with the team.

**Comment by pkocandr** [comment]
> Aside - I presume you would also be doing a matched PR for the documentation?

Right. I need to take a look into it too. Thanks for reminding me.

File: pom.xml (hunk 2 above)
Comment (rnc) [reply pom.xml:81]: OK, we will need to discuss this because right now it didn't even get to the tests as it isn't picking up the snapshot.

File: core/src/main/java/org/commonjava/maven/ext/core/state/RESTState.java [hunk 3]
```diff
@@ -84,6 +92,8 @@ public void initialise( Properties userProps )
         restSuffixAlign = Boolean.parseBoolean( userProps.getProperty( REST_SUFFIX, "true" ) );
 
         String repositoryGroup = userProps.getProperty( REST_REPO_GROUP, "" );
+        Boolean brewPullActive = BooleanUtils.toBooleanObject( userProps.getProperty( REST_BREW_PULL_ACTIVE ) );
+        Boolean temporaryBuild = BooleanUtils.toBooleanObject( userProps.getProperty( REST_TEMPORARY_BUILD ) );
```
Comment (rnc) [inline core/src/main/java/org/commonjava/maven/ext/core/state/RESTState.java:96]: Why not use `Boolean.parseBoolean` rather than an external library?

File: core/src/main/java/org/commonjava/maven/ext/core/state/RESTState.java (hunk 1 above)
Comment (rnc) [reply core/src/main/java/org/commonjava/maven/ext/core/state/RESTState.java:108]: Ok makes sense. Couple of options I guess - I release 4.3 now, and then this becomes 4.4 with `repositoryGroup` removed or we keep it and release in with 4.3 but mark it as deprecated via the `ConfigValue` annotation )with appropriate updates in the documentation) and then remove it in future.

File: core/src/main/java/org/commonjava/maven/ext/core/state/RESTState.java (hunk 3 above)
Comment (pkocandr) [reply core/src/main/java/org/commonjava/maven/ext/core/state/RESTState.java:96]: Because the logic using the flags also need to know if the value was passed in or not (using BooleanUtils.toBooleanObject(null) results in null). There is no such method in Boolean class AFAIK.

Comment (rnc) [reply core/src/main/java/org/commonjava/maven/ext/core/state/RESTState.java:96]: Huh, is the boolean being used as a tri-state value ?

File: pom.xml [hunk 4]
```diff
@@ -78,6 +78,7 @@
 
     <mavenVersion>3.5.0</mavenVersion>
     <atlasVersion>0.17.1</atlasVersion>
+    <daVersion>2.1.0.Api-1</daVersion>
```
Comment (rnc) [inline pom.xml:81]: Wow, what a GAV!! :-)

File: core/src/main/java/org/commonjava/maven/ext/core/state/RESTState.java [hunk 5]
```diff
@@ -84,6 +92,8 @@ public void initialise( Properties userProps )
         restSuffixAlign = Boolean.parseBoolean( userProps.getProperty( REST_SUFFIX, "true" ) );
 
         String repositoryGroup = userProps.getProperty( REST_REPO_GROUP, "" );
+        Boolean brewPullActive = BooleanUtils.toBooleanObject( userProps.getProperty( REST_BREW_PULL_ACTIVE ) );
```
Comment (rnc) [inline core/src/main/java/org/commonjava/maven/ext/core/state/RESTState.java:95]: Why can't we use `Boolean.parseBoolean` here?

File: io/src/main/java/org/commonjava/maven/ext/io/rest/DefaultTranslator.java [hunk 6]
```diff
@@ -101,15 +105,21 @@
      * @param endpointUrl is the URL to talk to.
      * @param restMaxSize initial (maximum) size of the rest call; if zero will send everything.
      * @param restMinSize minimum size for the call
-     * @param repositoryGroup the group to pass to the endpoint.
+     * @param repositoryGroup the group to pass to the endpoint, still can be used, but is replaced by
+     *        brewPullActive flag and mode when used in combination with DA 2.1 and above
+     * @param brewPullActive flag saying if brew pull should be used for version retrieval
```
Comment (rnc) [inline io/src/main/java/org/commonjava/maven/ext/io/rest/DefaultTranslator.java:110]: So just to clarify, `repositoryGroup` isn't deprecated as such - as its still in use, until the next PNC release when it can be removed? (i.e. we'll need a separate ticket to remind us to remove the flag for that?)

File: io/src/main/java/org/commonjava/maven/ext/io/rest/DefaultTranslator.java [hunk 7]
```diff
@@ -101,15 +105,21 @@
...
+     * @param mode lookup mode, either STANDARD (default if empty) or SERVICE
```
Comment (rnc) [inline io/src/main/java/org/commonjava/maven/ext/io/rest/DefaultTranslator.java:111]: What happens if someone passes something else (e.g. `foobar`) to the flag? Should it be an enum / or have some other checks? Or is there something in DA we can call on to validate it?

Comment (rnc) [reply io/src/main/java/org/commonjava/maven/ext/io/rest/DefaultTranslator.java:111]: Also what does it do? Should we have test to validate it?

File: core/src/main/java/org/commonjava/maven/ext/core/state/RESTState.java (hunk 1 above)
Comment (rnc) [reply core/src/main/java/org/commonjava/maven/ext/core/state/RESTState.java]: Marking as resolved - seems sensible to have a release with backwards compatibility

File: pom.xml (hunk 4 above)
Comment (pkocandr) [reply pom.xml:81]: Yeah, we needed to make a partial release because of a snapshot transitive dependency we cannot release ATM, so to emphisize that we chose this version suffix.

File: core/src/main/java/org/commonjava/maven/ext/core/state/RESTState.java (hunk 5 above)
Comment (pkocandr) [reply core/src/main/java/org/commonjava/maven/ext/core/state/RESTState.java:95]: Boolean.parseBoolean produces boolean, not Boolean, i.e. the null value (when not provided on command line) gets lost in favor of a default one. But the default should be decided on DA side, not in PME.

File: io/src/main/java/org/commonjava/maven/ext/io/rest/DefaultTranslator.java (hunk 6 above)
Comment (pkocandr) [reply io/src/main/java/org/commonjava/maven/ext/io/rest/DefaultTranslator.java:110]: Yeah, we introduced `brewPullActive` and `mode` which tells DA which repository group to use (if any). The `repositoryGroup` parameter is meant to be used in combination with DA 2.0 and with DA 2.1 we will switch to use the other 2.

File: io/src/main/java/org/commonjava/maven/ext/io/rest/DefaultTranslator.java (hunk 7 above)
Comment (pkocandr) [reply io/src/main/java/org/commonjava/maven/ext/io/rest/DefaultTranslator.java:111]: The modes are dynamically configured in DA as needed.  Currently we plan to use 4 - PERSISTENT, TEMPORARY, SERVICE and SERVICE-TEMPORARY. It specifies allowed version suffixes, build categories and artifact qualities when looking up existing versions. If somebody passes a non-existing mode, it will fail in DA.

File: pom.xml (hunk 4 above)
Comment (rnc) [reply pom.xml:81]: Ok!

File: io/src/main/java/org/commonjava/maven/ext/io/rest/DefaultTranslator.java (hunk 6 above)
Comment (rnc) [reply io/src/main/java/org/commonjava/maven/ext/io/rest/DefaultTranslator.java:110]: Ok thanks! Can another ticket be spawned to note removal of `repositoryGroup` please?

File: core/src/main/java/org/commonjava/maven/ext/core/state/RESTState.java (hunk 5 above)
Comment (rnc) [reply core/src/main/java/org/commonjava/maven/ext/core/state/RESTState.java:95]: Won't it get serialized via jackson when calling DA? So it would end up as false/true ?

File: io/src/main/java/org/commonjava/maven/ext/io/rest/DefaultTranslator.java (hunk 7 above)
Comment (rnc) [reply io/src/main/java/org/commonjava/maven/ext/io/rest/DefaultTranslator.java:111]: So, inside PNC would repour pass the correct mode to pme, which passes it through to DA?

And the mode isn't hardcoded in DA, but configured on service start? Should pme handle non-existing mode specially ?

File: core/src/main/java/org/commonjava/maven/ext/core/state/RESTState.java (hunk 5 above)
Comment (pkocandr) [reply core/src/main/java/org/commonjava/maven/ext/core/state/RESTState.java:95]: It depends on the serialization implementation, but I expected that null value will be skipped from putting into the resulting json.
But anyway, the exptra fields are ignored and in DA 2.1 these become mandatory, so sending in a value is probably ok. Will update it to use `Boolean.parseBoolean()`.

File: io/src/main/java/org/commonjava/maven/ext/io/rest/DefaultTranslator.java (hunk 6 above)
Comment (pkocandr) [reply io/src/main/java/org/commonjava/maven/ext/io/rest/DefaultTranslator.java:110]: Created as NCL-6528. Thank you!

File: io/src/main/java/org/commonjava/maven/ext/io/rest/DefaultTranslator.java (hunk 7 above)
Comment (pkocandr) [reply io/src/main/java/org/commonjava/maven/ext/io/rest/DefaultTranslator.java:111]: Yes, Repour is responsible to pass the mode based on the build type etc. and the modes are configured in both DA and Repour at service start.
If a mode doesn't exist it should be returned by DA in an error response. If used directly by users they should be able to read it from the log (I expect the exception from the response should be printed in the terminal) and in case of running in PNC the sustaining team will find it in the logs to understand the problem and fix the config.
What kind of special handling would you suggest?

Comment (rnc) [reply io/src/main/java/org/commonjava/maven/ext/io/rest/DefaultTranslator.java:111]: None really :-) Let me rephrase, if an erroneous response is printed sufficiently clearly for people to understand, thats fine. I was just wondering if the error response should be parsed and in this case made really explicit what has gone wrong. But if you think its fine as is, thats fine.

Comment (pkocandr) [reply io/src/main/java/org/commonjava/maven/ext/io/rest/DefaultTranslator.java:111]: Actually when I tested it now it failed with `REST communication with http://da-master-devel.psi.redhat.com/da/rest/v-1 failed. Received response status -1 with message: Problem in HTTP communication with status code 500 and message Internal Server Error`. This is definitely suboptimal, but must be handled on DA side IMO.

**Review by rnc — APPROVED** [review]

File: io/src/main/java/org/commonjava/maven/ext/io/rest/DefaultTranslator.java (hunk 7 above)
Comment (rnc) [reply io/src/main/java/org/commonjava/maven/ext/io/rest/DefaultTranslator.java:111]: Sure, we can always do another iteration I guess if PME needs further changes

Comment (pkocandr) [reply io/src/main/java/org/commonjava/maven/ext/io/rest/DefaultTranslator.java:111]: I've added one more commit to handle this. It turned out the exception is in the response body, but PME did not print it even with debug on. Now it is printed like this:
```ERROR org.commonjava.maven.ext.cli.Cli - REST communication with http://da.example.com/da/rest/v-1 failed. Received response status 500 with message: javax.validation.ValidationException: Invalid mode name: a. Available modes are: [PERSISTENT, TEMPORARY]```
